### PR TITLE
NN-4114: Caching of auth client credentials

### PR DESCRIPTION
**Pull request by sp-watson**

### CODE REVIEW

File: backend/api/systemOauthClient.ts [hunk 1]
```diff
@@ -1,31 +1,87 @@
...
-      url: config.apis.oauth2.url,
-    }
-  ).makeTokenRequest(oauthRequest, 'PSH-client_credentials')
+  const oauthResult = await oauthClient.makeTokenRequest(oauthRequest, 'PSH-client_credentials')
 
   logger.debug(`Oauth request for grant type 'client_credentials', result status: successful`)
   return oauthResult
 }
 
+const debug = (operation: string, username: string) => {
```
Comment (sp-watson) [inline backend/api/systemOauthClient.ts:49]: We ought to remove this once we are confident it is working OK

File: backend/index.ts [hunk 2]
```diff
@@ -48,6 +49,7 @@ app.set('view engine', 'njk')
 
 nunjucksSetup(app)
 phaseNameSetup(app, config)
+clientCredsSetup(getTokenStore(config), apis.oauthApi)
```
Comment (sp-watson) [inline backend/index.ts:52]: I had to add this setup code to facilitate testing

File: backend/api/systemOauthClient.ts [hunk 3]
```diff
@@ -1,31 +1,87 @@
...
+  if (!enabled || !host) return null
+
+  const client = createClient({
+    host,
+    port,
+    password,
+    tls: config.app.production ? {} : false,
+  })
+
+  client.on('error', (e: Error) => logger.error('Redis client error', e))
```
Comment (steverendell) [inline backend/api/systemOauthClient.ts:22]: Doesnt the e parameter come first?

Comment (sp-watson) [reply backend/api/systemOauthClient.ts:22]: You mean in the `error` function parameters? Hmm.. yes.. which means the other 2 places in DPS are wrong as well. I'll fix them

Good spot!

**Review by timoj58 — APPROVED** [review]

**Pull request merged**